src/item.h

```diff
@@ -206,19 +210,271 @@ class ItemAttributes
+				}
+
+				void operator()(const int64_t& v) const {
+					propWriteStream.write<int64_t>(v);
+				}
+			};
+
+			void serialize(PropWriteStream& propWriteStream) const {
+				propWriteStream.write<uint8_t>(static_cast<uint8_t>(value.which()));
+				boost::apply_visitor(serializeVisitor(propWriteStream), value);
+			}
+
+			bool unserialize(PropStream& propStream) {
+				// This is hard coded so it's not general, depends on the position of the variants.
+				uint8_t pos;
+				if (!propStream.read<uint8_t>(pos)) {
+					return false;
+				}
+
+				switch (pos) {
+					case 1:  { // std::string
+						std::string tmp;
+						if (!propStream.readString(tmp)) {
+							return false;
+						}
+						value = tmp;
+						break;
+					}
+
+					case 2: { // int64_t
+						int64_t tmp;
+						if (!propStream.read<int64_t>(tmp)) {
+							return false;
+						}
+						value = tmp;
+						break;
+					}
+
+					default: {
+						value = boost::blank();
+						return false;
+					}
+				}
+				return true;
+			}
+
+			struct equalityVisitor : public boost::static_visitor<bool> {
+				bool operator()(const std::string& v1, const std::string& v2) const {
+					return !v1.compare(v2);
+				}
+
+				bool operator()(const int64_t& v1, const int64_t& v2) const {
+					return v1 == v2;
+				}
+
+				template<class T1, class T2>
+				bool operator()(const T1& v1, const T2& v2) const {
+					return false;
+				}
+			};
+
+			bool operator==(const CustomAttributeKey& other) const {
+				if (value.which() != other.value.which()) {
+					return false;
+				}
+
+				return boost::apply_visitor(equalityVisitor(), value, other.value);
+			}
+
+			struct hashVisitor : public boost::static_visitor<size_t> {
+				size_t operator()(const boost::blank& value) const {
+					return 0;
+				}
+
+				template<typename T>
+				size_t operator()(const T& value) const {
+					return std::hash<T>()(value);
+				}
+			};
+
+			struct hash {
+				std::size_t operator()(const CustomAttributeKey& key) const {
+					return boost::apply_visitor(hashVisitor(), key.value);
+				}
+			};
+		};
+
+		struct CustomAttribute
+		{
+			typedef boost::variant<boost::blank, std::string, int64_t, double, bool> VariantAttribute;
+			VariantAttribute value;
+
+			CustomAttribute() : value(boost::blank()) {}
+
+			template<typename T>
+			explicit CustomAttribute(const T& v) : value(v) {}
+
+			template<typename T>
+			void set(const T& v) {
+				value = v;
+			}
+
+			template<typename T>
+			const T get() {
+				if (value.type() == typeid(T)) {
+					return boost::get<T>(value);
+				}
+				return T();
+			}
+
+			struct pushLuaVisitor : public boost::static_visitor<> {
+				lua_State* L;
+
+				pushLuaVisitor(lua_State* L) : boost::static_visitor<>(), L(L) {}
+
+				void operator()(const boost::blank& v) const {
+					lua_pushnil(L);
+				}
+
+				void operator()(const std::string& v) const {
+					LuaScriptInterface::pushString(L, v);
+				}
+
+				void operator()(const bool& v) const {
```

Comment:
Fundamental types can be passed by value - they usually fit in registers. But I think it won't matter here, because these functions will be inlined anyway.